### DISCUSSION TITLE
More EnumerateLibrary woes - removal of unwanted hydrogens

### DISCUSSION THREAD

**Discussion by mark-mackey-cresset** [rdkit/rdkit#6116] [Q&A]
Hi all,

I'm having trouble getting the EnumerateLibrary function to behave itself when converting atoms from sp3 to sp2. This originally cropped up for the Paul-Knorr pyrrole synthesis, but it's easier to see in a simpler example:

```
from rdkit import Chem
from rdkit.Chem.rdChemReactions import ReactionFromSmarts, EnumerateLibrary

# Aromatisation reaction
rxn = r"[C:1]1[C:2][C:3][C:4][C:5][C:6]1>>[c:1]1:[c:2]:[c:3]:[c:4]:[c:5]:[c:6]:1"
r1s=["C1[C@H](C)C[C@@H](C)CC1"]

rxn = ReactionFromSmarts(rxn)
bbs=[[Chem.MolFromSmiles(x) for x in r1s]]

enumerator = EnumerateLibrary(rxn, bbs)
for prods in enumerator:
    for mols in prods:
        for mol in mols:
            print("Product", Chem.MolToSmiles(mol))
```

This prints out "`C[c@H]1ccc[c@H](C)c1`" which is obviously not what was intended. What am I missing?

**Comment by greglandrum** [comment]
Hi @mark-cresset, this feels like a bug to me. 
The reaction code should probably recognize that the atoms which were previously chiral but which have lost an H atom are no longer chiral. Either that or we need to provide a function which allows you to recover from this.
In the meantime the only real fix is a pretty painful bit of code to recognize those centers and manually remove the stereo.
If you're interested I can put something together for that and share it.

**Reply by mark-mackey-cresset** [reply]
Hi Greg, thanks for the quick reply. Would removing the stereo flags on the products also get rid of the extraneous hydrogens, or would I also have to fiddle around with the implicit H count or something similar?

**Reply by mark-mackey-cresset** [reply]
Hi Greg. A few pointers on how to clean up the product molecules would be appreciated - I'm not really getting anywhere on this, unfortunately.

**Reply by greglandrum** [reply]
Hi @mark-cresset , the quickest solution to this that I can come up with requires the unpleasant step of adding and removing Hs from reactants/products but does accurately reflect the chemistry in your system and produce a correct result:
```
rxn = r"[H][C:1]1[C:2]([H])[C:3]([H])[C:4]([H])[C:5]([H])[C:6]1[H]>>[c:1]1:[c:2]:[c:3]:[c:4]:[c:5]:[c:6]:1"
r1s=["C1[C@H](C)C[C@@H](C)CC1"]

rxn = ReactionFromSmarts(rxn)
bbs=[[Chem.AddHs(Chem.MolFromSmiles(x)) for x in r1s]]

enumerator = EnumerateLibrary(rxn, bbs)
for prods in enumerator:
    for mols in prods:
        for mol in mols:
            print("Product", Chem.MolToSmiles(Chem.RemoveHs(mol)))
```
Including some indication that there have to be H atoms in the reactants is not a bad idea for this reaction anyway, it shouldn't apply to something like `C1CCCCC1(C)(O)`

**Reply by mark-mackey-cresset** [reply]
Hmmm, that's not going to work for us as adding explicit Hs to the reactants breaks other reactions. Could I instead clean this up by iterating over the atoms in the product, finding atoms that have a tetrahedral stereochemistry set but appear to be sp2, clearing the stereochemistry flag and reducing the implicit H count by 1? If so, is there an easy way in the C++ API to fiddle with the implicit H count?

**Comment by ptosco** [comment]
@mark-cresset This should do what you need:
```
def has_double_or_aromatic_bond(a):
    return any(b.GetIsAromatic()
        or b.GetBondType() == Chem.BondType.DOUBLE
        for b in a.GetBonds())

def remove_stereo_from_sp2(mol):
    for a in mol.GetAtoms():
        if (a.GetAtomicNum() < 11
            and has_double_or_aromatic_bond(a)
            and a.GetChiralTag() != Chem.ChiralType.CHI_UNSPECIFIED):
            a.SetChiralTag(Chem.ChiralType.CHI_UNSPECIFIED)
            a.SetNumExplicitHs(max(0, a.GetNumExplicitHs() - 1))
    Chem.SanitizeMol(mol)

# Aromatisation reaction
rxn = r"[C!H0:1]1[C!H0:2][C!H0:3][C!H0:4][C!H0:5][C!H0:6]1>>[c:1]1:[c:2]:[c:3]:[c:4]:[c:5]:[c:6]:1"
r1s = ["C1[C@H](C)C[C@@H](C)CC1"]
rxn = ReactionFromSmarts(rxn)
bbs = [[Chem.MolFromSmiles(x) for x in r1s]]
enumerator = EnumerateLibrary(rxn, bbs)
for prods in enumerator:
    for mols in prods:
        for mol in mols:
            remove_stereo_from_aromatic(mol)
            print("Product", Chem.MolToSmiles(mol))
Product Cc1cccc(C)c1
Product Cc1cccc(C)c1
Product Cc1cccc(C)c1
Product Cc1cccc(C)c1
Product Cc1cccc(C)c1
Product Cc1cccc(C)c1
Product Cc1cccc(C)c1
Product Cc1cccc(C)c1
Product Cc1cccc(C)c1
Product Cc1cccc(C)c1
Product Cc1cccc(C)c1
Product Cc1cccc(C)c1

# Introduce an unsaturation
rxn = r"[C!H0:1]1-[C!H0:2]CCCC1>>[C:1]1=[C:2]CCCC1"
r1s = ["C1[C@H](C)C[C@@H](C)CC1"]
bbs = [[Chem.MolFromSmiles(x) for x in r1s]]
rxn = ReactionFromSmarts(rxn)
enumerator = EnumerateLibrary(rxn, bbs)
for prods in enumerator:
    for mols in prods:
        for mol in mols:
            remove_stereo_from_sp2(mol)
            print("Product", Chem.MolToSmiles(mol))
Product CC1=CCCCC1
Product C1=CCCCC1
Product CC1=CCCCC1
Product CC1=CCCCC1
Product CC1=CCCCC1
Product CC1=CCCCC1
Product CC1=CCCCC1
Product CC1=CCCCC1
Product CC1=CCCCC1
Product C1=CCCCC1
Product C1=CCCCC1
Product C1=CCCCC1
```
I think you should specify that former _sp3_ centres that undergo unsaturation should have had at least one hydrogen to start with and I modified your reaction SMARTS accordingly.

**Reply by mark-mackey-cresset** [reply]
Excellent - thanks Paolo. Just added this in and it works well.
